app/connectors/HipConnector.scala

```diff
@@ -0,0 +1,237 @@
+      case e: UpstreamErrorResponse if e.statusCode >= 500 && e.statusCode < 600 => true
+      case _ => false
+    }
+  }
+
+  // Returns Right(success) for 200 and Left(failures) for 422
+  def calculateOutcome(userId: String, request: HipCalculationRequest)(implicit hc: HeaderCarrier): Future[Either[HipCalculationFailuresResponse, HipCalculationResponse]] = {
+    doAudit(
+      "gmpCalculation",
+      userId,
+      request.schemeContractedOutNumber,
+      Some(request.nationalInsuranceNumber),
+      Some(request.surname),
+      Some(request.firstForename)
+    )
+
+    val startTime = System.currentTimeMillis()
+    val headers = buildHeadersV1
+
+    withCircuitBreaker(
+      http.post(url"$calcURI")
+        .setHeader(headers*)
+        .withBody(Json.toJson(request))
+        .execute[HttpResponse]
+        .map[Either[HipCalculationFailuresResponse, HipCalculationResponse]] { response =>
+          val took = System.currentTimeMillis() - startTime
+          metrics.hipConnectionTime(took, TimeUnit.MILLISECONDS)
+          metrics.hipRegisterStatusCode(response.status.toString)
+
+          val cid = response.headers.get("correlationId").flatMap(_.headOption).getOrElse("n/a")
+          logger.debug(s"[HipConnector][calculate] Response status: ${response.status}, correlationId: $cid")
+
+          response.status match {
+            case OK =>
+              metrics.hipRegisterSuccessfulRequest()
+              Right(parseHipCalculationSuccess(response))
+            case Status.UNPROCESSABLE_ENTITY =>
+              metrics.hipRegisterSuccessfulRequest()
+              Left(parseHipFailures(response))
+            case _ =>
+              metrics.hipRegisterFailedRequest()
+              throw toUpstreamError(response)
+          }
+        }
+    )
+  }
+
+  private def parseHipCalculationSuccess(response: HttpResponse): HipCalculationResponse = {
+    logger.debug(s"[HipConnector][calculate] Response body: ${LoggingUtils.redactCalculationData(response.body)}")
+
+    scala.util.Try(Json.parse(response.body)).toOption match {
+      case Some(js) =>
+        js.validate[HipCalculationResponse] match {
+          case JsSuccess(value, _) =>
+            logger.debug(s"[HipConnector] Successfully parsed HIP calculation response")
+            value
+          case JsError(errors) =>
+            val errorFields = errors.map(_._1.toString).mkString(", ")
+            val detailedMsg = s"HIP returned invalid JSON (status: ${response.status}). Failed to parse fields: $errorFields"
+            logger.error(s"[HipConnector][calculate] $detailedMsg")
+            metrics.hipRegisterFailedRequest()
+            throw UpstreamErrorResponse(detailedMsg, BAD_GATEWAY, BAD_GATEWAY)
+        }
+      case None =>
+        val detailedMsg = s"HIP returned non-JSON body with ${response.status}. Body: ${response.body.take(MaxBodyLengthForLogging)}"
+        logger.error(s"[HipConnector][calculate] $detailedMsg")
```

Comment:
should we be logging the body? this is a case for parsing failures and it would result from invalid spaces in json but still contain sensitive data from response. So, by logging here we are risking getting sensitive data into logs. These should be audited for resolving any incidents.